File: frontend/src/services/api/planner.ts
import { apiClient } from './client';
import { Task, Priority } from '@/types';

export interface PlannerEntry {
  id: string;
  user_id: string;
  task_id: string;
  start_at: string;
  end_at: string;
  status: 'SCHEDULED' | 'COMPLETED' | 'MISSED' | 'CANCELLED';
  source: 'MANUAL' | 'AUTO_SUGGESTED' | 'AI_COACH';
  created_at: string;
  updated_at: string;
  task?: Task;
}

export interface CreatePlannerEntryInput {
  task_id: string;
  start_at: string;
  end_at: string;
  status?: string;
  source?: string;
}

export interface AutoScheduleSuggestion {
  task_id: string;
  task_title: string;
  start_at: string;
  end_at: string;
  priority: Priority;
}

export interface AutoScheduleResponse {
  date: string;
  available_hours: number;
  total_planned_hours: number;
  is_overloaded: boolean;
  suggestions: AutoScheduleSuggestion[];
}

export const plannerApi = {
  list: (startTime?: string, endTime?: string) => {
    const params = new URLSearchParams();
    if (startTime) params.append('start_time', startTime);
    if (endTime) params.append('end_time', endTime);
    const qs = params.toString();
    return apiClient<PlannerEntry[]>(qs ? `/planner?${qs}` : '/planner');
  },

  create: (data: CreatePlannerEntryInput) =>
    apiClient<PlannerEntry>('/planner', {
      method: 'POST',
      body: JSON.stringify(data),
    }),

  update: (id: string, data: Partial<CreatePlannerEntryInput>) =>
    apiClient<PlannerEntry>(`/planner/${id}`, {
      method: 'PATCH',
      body: JSON.stringify(data),
    }),

  delete: (id: string) =>
    apiClient<{ message: string }>(`/planner/${id}`, {
      method: 'DELETE',
    }),

  autoSchedule: (date?: string) =>
    apiClient<AutoScheduleResponse>('/planner/auto-schedule', {
      method: 'POST',
      body: JSON.stringify({ schedule_date: date }),
    }),
};

File: backend/app/schemas/planner.py
from pydantic import BaseModel, ConfigDict
from uuid import UUID
from datetime import datetime
from typing import Optional, List, Any
from app.schemas.task import TaskResponse

class PlannerEntryBase(BaseModel):
    task_id: UUID
    start_at: datetime
    end_at: datetime
    status: str = "SCHEDULED" # SCHEDULED, COMPLETED, MISSED, CANCELLED
    source: str = "MANUAL"    # MANUAL, AUTO_SUGGESTED, AI_COACH

class PlannerEntryCreate(PlannerEntryBase):
    pass

class PlannerEntryUpdate(BaseModel):
    start_at: Optional[datetime] = None
    end_at: Optional[datetime] = None
    status: Optional[str] = None

class PlannerEntryResponse(PlannerEntryBase):
    id: UUID
    user_id: UUID
    created_at: datetime
    updated_at: datetime
    task: Optional[TaskResponse] = None

    model_config = ConfigDict(from_attributes=True)

class AutoScheduleRequest(BaseModel):
    schedule_date: Optional[str] = None

class AutoScheduleSuggestion(BaseModel):
    task_id: UUID
    task_title: str
    start_at: datetime
    end_at: datetime
    priority: str

class AutoScheduleResponse(BaseModel):
    date: str
    available_hours: float
    total_planned_hours: float
    is_overloaded: bool
    suggestions: List[AutoScheduleSuggestion]

class ScheduleSlotProposal(BaseModel):
    task_id: Any
    task_title: str
    priority: str
    estimated_minutes: Optional[int] = 30
    proposed_start_time: Any
    proposed_end_time: Any
    reasoning: Optional[str] = None

class PlannerSuggestionResponse(BaseModel):
    total_suggested_minutes: int
    slots: List[Any]

class BatchConfirmScheduleRequest(BaseModel):
    confirmed_slots: List[Any]
